bootspec/src/v1.rs

```diff
@@ -37,11 +37,12 @@ pub struct GenerationV1 {
     /// Path to "append-initrd-secrets" script -- $toplevel/append-initrd-secrets
     pub initrd_secrets: Option<PathBuf>,
     /// Mapping of specialisation names to their boot.json
-    pub specialisation: HashMap<SpecialisationName, GenerationV1>,
+    pub specialisation: HashMap<SpecialisationName, GenerationV1<Extension>>,
     /// config.system.build.toplevel path
     pub toplevel: SystemConfigurationRoot,
     /// User extensions for this specification
-    pub extensions: HashMap<String, serde_json::Value>
+    #[serde(default)]
```

Comment:
Does this `default` let serde skip the field if it doesn't exist? If not, I feel like there should be some way to do that (maybe just by making `extensions` an `Option`?)

Reply:
@RaitoBezarius I experimented with removing the Default trait bound and have some code lying around in my fork for that: https://github.com/nikstur/bootspec-1/tree/secureboot-needs. You're welcome to use this if you want to. My fork also addresses my other remarks (TryFrom instead of TryInto and non exhaustive matching)

Reply:
The problem of having an `Option` I guess is that: it does not lend naturally in the case of HashMap<X, Y> where the natural "none" is an empty HashMap

Reply:
Hence, my proposal of TryDefault

Reply:
- We may want to have some fields as Default
- but not all of them
- etc.

Reply:
Okay, I thought hard and what I want is:
```
trait TryDefaultDeserialize<'de, D: Deserializer<'de>, T: Deserialize<'de>> {
    fn deserialize(deserializer: D) -> Result<T, D::Error>;
}

impl<'de, D: Deserializer<'de>, Extension: Deserialize<'de>> TryDefaultDeserialize<'de, D, Extension> for D {
    default fn deserialize(deserializer: D) -> Result<Extension, D::Error> {
        Extension::deserialize(deserializer)
    }
}

impl<'de, D: Deserializer<'de>, Extension: Deserialize<'de> + Default> TryDefaultDeserialize<'de, D, Extension> for D {
    fn deserialize(deserializer: D) -> Result<Extension, D::Error>
    {
        Ok(Extension::deserialize(deserializer).unwrap_or_else(|_| Extension::default()))
    }
}
```

Unfortunately, this is an unstable feature of Rust at the moment, see https://github.com/rust-lang/rust/issues/31844

Reply:
So, I'm fine with Option for now, because we are not going to get something better until this RFC lands in stable

Reply:
But it do mean that deserialization of a missing `extensions` field is not going to fail, unfortunately, and users will have to take care of this themselves.

Reply:
I think that's good, because the `extensions` field is not (currently) a part of the RFC text except as an open question. If it becomes a part of the implementation described in the RFC text itself, then maybe we can revisit this and make it non-`Option`al.

Reply:
Last time we discussed the RFC, we decided to make them part of the RFC AFAIK.

Reply:
see https://github.com/NixOS/rfcs/pull/125#issuecomment-1287823541 (that was during NixCon)

Reply:
> The problem of having an `Option` I guess is that: it does not lend naturally in the case of HashMap<X, Y> where the natural "none" is an empty HashMap

I think this mostly depends on the language you use the most. I'd wager that the natural `None` for someone that regularly uses is actually an explicit `Option` instead of an empty hash map.

Another perspective is this: whether you get an empty hash map or an `Option` you still have to handle the possibility that the field doesn't exist. The explicit `Option` makes this clear and Rust also provides nice facilities to deal with that.

Reply:
Right, I have this FP/Haskell syndrome which leads me to pursue for some Traversable object in this situation.
The annoying problem in this situation is that: in the RFC, we said that optional fields can either be:

- an actual value of the type T, i.e. an empty hashmap is valid
- *absence* of the field

But *not* a null-value, i.e. `null` in JSON.

Having `Option<T>` is allowing `null` as valid values for bootspec. It's not a big deal as our Nix generator uses Cue for schema validation, but theoretically, the Bootspec library should not deserialize any nulls for optional values.

Hence, why I am pushing for some sort of TryDefault which is closest thing to what would make it good according to the specification.

Now, I am seeing this is hard to maintain, so I'm not sure, we want to keep this constraint in the RFC.
The biggest problem I see about introducing null as valid optional field is that if some value can be either: Null Or T, null *becomes* de-facto a valid value, and we get ambiguity between optional field and valid values.

So I wished that serde would not deserialize null in case of Option<T> and reject them nevertheless.

Reply:
I will push additional tests to reflect this problem.

Reply:
this https://github.com/DeterminateSystems/bootspec/pull/68/commits/1587739421e35e6b789b0763e2d11692d352496f#diff-3598be44e289e7ea3af0b580c43ec71823b33addb89c6542481f4acfc1fcc6f6R271 should be a failing test according to the RFC IMHO

Reply:
See https://github.com/serde-rs/serde/issues/984

Reply:
I understand that that's academically frustrating, but what would be the real world problem of accepting `null` as well as the absence of a field?

Reply:
> I understand that that's academically frustrating, but what would be the real world problem of accepting `null` as well as the absence of a field?

ambiguity: `extensions: Option<Option<i32>>` how do you express `Some(None)` ?

Reply:
(I'm still thinking of a real world problem)

Reply:
Null coercion to a value is always ambiguous and could be used to vehicle the meaning of an actual value and not `None`.

For example: 

- `Option<String>`: `null` → `Some("")` or `None` could be two valid deserializations.
- `Option<HashMap<String, serde_json::Value>>` could be deserialized to `HashMap::new()`
- `Option<{ osRelease: Option<String> }>` could be deserialized from `{ }` (this is the case here, but it's not obvious to reason about!), what about deserialization from `null`, should that be the same? IMHO, yes. But due to semantics™, it will provide a `None` value.

I will provide a new test for the last one, that's the reason why I want to have absence for "optional" field at the top-level.

Reply:
https://github.com/DeterminateSystems/bootspec/pull/68/commits/a123d9b71a4c72e832309de43b49e6b4defdee02 takes a lot simpler approach when we discussed it with @lheckemann so I resolve this thread, don't hesitate to open another one